### DISCUSSION TITLE
Unable to load fine-tuned transformers model

### DISCUSSION THREAD

**Discussion by DNau27** [explosion/spaCy#8717] [Help: Coding & Implementations] [answered]
Hi, 
I'm trying to load transformer model's I've trained strongly leaning on the project template on https://github.com/explosion/projects/tree/v3/tutorials/textcat_docs_issues. 

When trying to load the model in a Jupyter Notebook, I get the following error message, even though the loaded model should be stored locally. Can these models trained on command line only be accessed from command line, as well?

(I don't have an internet connection on my company machine so I had to install what was needed to train the models from file.)

Thanks in advance!

`ValueError                                Traceback (most recent call last)
<ipython-input-23-cd814b72088b> in <module>
      1 import spacy
----> 2 nlp = spacy.load("model-best/transformer")

~/.conda/envs/dev_spacy_3/lib/python3.6/site-packages/spacy/__init__.py in load(name, vocab, disable, exclude, config)
     49     """
     50     return util.load_model(
---> 51         name, vocab=vocab, disable=disable, exclude=exclude, config=config
     52     )
     53 

~/.conda/envs/dev_spacy_3/lib/python3.6/site-packages/spacy/util.py in load_model(name, vocab, disable, exclude, config)
    324             return load_model_from_package(name, **kwargs)
    325         if Path(name).exists():  # path to model data directory
--> 326             return load_model_from_path(Path(name), **kwargs)
    327     elif hasattr(name, "exists"):  # Path or Path-like to model data
    328         return load_model_from_path(name, **kwargs)

~/.conda/envs/dev_spacy_3/lib/python3.6/site-packages/spacy/util.py in load_model_from_path(model_path, meta, vocab, disable, exclude, config)
    390     config = load_config(config_path, overrides=dict_to_dot(config))
    391     nlp = load_model_from_config(config, vocab=vocab, disable=disable, exclude=exclude)
--> 392     return nlp.from_disk(model_path, exclude=exclude)
    393 
    394 

~/.conda/envs/dev_spacy_3/lib/python3.6/site-packages/spacy/language.py in from_disk(self, path, exclude)
   1881             # Convert to list here in case exclude is (default) tuple
   1882             exclude = list(exclude) + ["vocab"]
-> 1883         util.from_disk(path, deserializers, exclude)
   1884         self._path = path
   1885         self._link_components()

~/.conda/envs/dev_spacy_3/lib/python3.6/site-packages/spacy/util.py in from_disk(path, readers, exclude)
   1174         # Split to support file names like meta.json
   1175         if key.split(".")[0] not in exclude:
-> 1176             reader(path / key)
   1177     return path
   1178 

~/.conda/envs/dev_spacy_3/lib/python3.6/site-packages/spacy/language.py in <lambda>(p, proc)
   1876                 continue
   1877             deserializers[name] = lambda p, proc=proc: proc.from_disk(
-> 1878                 p, exclude=["vocab"]
   1879             )
   1880         if not (path / "vocab").exists() and "vocab" not in exclude:

~/.conda/envs/dev_spacy_3/lib/python3.6/site-packages/spacy_transformers/pipeline_component.py in from_disk(self, path, exclude)
    400             "model": load_model,
    401         }
--> 402         util.from_disk(path, deserialize, exclude)
    403         return self

~/.conda/envs/dev_spacy_3/lib/python3.6/site-packages/spacy/util.py in from_disk(path, readers, exclude)
   1174         # Split to support file names like meta.json
   1175         if key.split(".")[0] not in exclude:
-> 1176             reader(path / key)
   1177     return path
   1178 

~/.conda/envs/dev_spacy_3/lib/python3.6/site-packages/spacy_transformers/pipeline_component.py in load_model(p)
    390             p = Path(p).absolute()
    391             tokenizer, transformer = huggingface_from_pretrained(
--> 392                 p, self.model.attrs["tokenizer_config"]
    393             )
    394             self.model.attrs["tokenizer"] = tokenizer

~/.conda/envs/dev_spacy_3/lib/python3.6/site-packages/spacy_transformers/util.py in huggingface_from_pretrained(source, config)
     29     else:
     30         str_path = source
---> 31     tokenizer = AutoTokenizer.from_pretrained(str_path, **config)
     32     transformer = AutoModel.from_pretrained(str_path)
     33     ops = get_current_ops()

~/.conda/envs/dev_spacy_3/lib/python3.6/site-packages/transformers/models/auto/tokenization_auto.py in from_pretrained(cls, pretrained_model_name_or_path, *inputs, **kwargs)
    388         kwargs["_from_auto"] = True
    389         if not isinstance(config, PretrainedConfig):
--> 390             config = AutoConfig.from_pretrained(pretrained_model_name_or_path, **kwargs)
    391 
    392         use_fast = kwargs.pop("use_fast", True)

~/.conda/envs/dev_spacy_3/lib/python3.6/site-packages/transformers/models/auto/configuration_auto.py in from_pretrained(cls, pretrained_model_name_or_path, **kwargs)
    396         """
    397         kwargs["_from_auto"] = True
--> 398         config_dict, _ = PretrainedConfig.get_config_dict(pretrained_model_name_or_path, **kwargs)
    399         if "model_type" in config_dict:
    400             config_class = CONFIG_MAPPING[config_dict["model_type"]]

~/.conda/envs/dev_spacy_3/lib/python3.6/site-packages/transformers/configuration_utils.py in get_config_dict(cls, pretrained_model_name_or_path, **kwargs)
    464                 local_files_only=local_files_only,
    465                 use_auth_token=use_auth_token,
--> 466                 user_agent=user_agent,
    467             )
    468             # Load config dict

~/.conda/envs/dev_spacy_3/lib/python3.6/site-packages/transformers/file_utils.py in cached_path(url_or_filename, cache_dir, force_download, proxies, resume_download, user_agent, extract_compressed_file, force_extract, use_auth_token, local_files_only)
   1171             user_agent=user_agent,
   1172             use_auth_token=use_auth_token,
-> 1173             local_files_only=local_files_only,
   1174         )
   1175     elif os.path.exists(url_or_filename):

~/.conda/envs/dev_spacy_3/lib/python3.6/site-packages/transformers/file_utils.py in get_from_cache(url, cache_dir, force_download, proxies, etag_timeout, resume_download, user_agent, use_auth_token, local_files_only)
   1387                 else:
   1388                     raise ValueError(
-> 1389                         "Connection error, and we cannot find the requested files in the cached path."
   1390                         " Please try again or make sure your Internet connection is on."
   1391                     )

ValueError: Connection error, and we cannot find the requested files in the cached path. Please try again or make sure your Internet connection is on.`

**Comment by BramVanroy** [comment, marked as answer]
There might be an easier way with [these methods](https://spacy.io/usage/saving-loading#pipeline) but I found that [packaging](https://spacy.io/api/cli#package) is easiest. It packages the whole pipeline into a pip-installable package, which you can then call like any other spaCy model. E.g.

```bash
python -m spacy package training/your-treebank/model-best packages
python -m pip install packages/your-package-name
```

and then in spaCy

```python
nlp = spacy.load("your-package-name")
```

**Reply by DNau27** [reply]
Hi @BramVanroy,
thanks a lot for your answer! 
I was able to create packages for my models and they can be loaded via 
`nlp = spacy.load("package-name")`
without any error messages arising. However, I can't access the specific functions of my models. They contain textcat and transformers. I'd expect the sentence to receive either a score or a category when calling 

```
import spacy
nlp = spacy.load("package-name")
doc123 = nlp("Example Sentence to be categorized.")
print(doc123)
```

but get

```
Example Sentence to be categorized.
```

so it simply returns the sentence I applied nlp to.
I'm pretty sure I'm missing a step there, but I can't seem to figure out what it is. Any help would be much appreciated!

**Reply by polm** [reply]
A spaCy pipeline (the `nlp` object) takes text and returns a Doc, and when you print a Doc it just shows the text. But other information can be stored in attributes on the Doc. In the case of a textcat pipeline the cats are in `doc.cats`, so if you `print(doc123.cats)` you should be able to see the predictions.

**Reply by DNau27** [reply]
Oh wow! Thanks @polm, how did I miss that?! :smile: 
Works like a charm now. Thanks again to @BramVanroy, as well!
